File: src/main/java/com/qtech/bubbles/event/bus/Bus.java
package com.qtech.bubbles.event.bus;

import com.qtech.bubbles.BubbleBlaster;
import com.qtech.bubbles.addon.loader.AddonContainer;
import com.qtech.bubbles.common.mod.ModInstance;
import com.qtech.bubbles.common.mod.ModObject;
import org.jetbrains.annotations.NotNull;

@SuppressWarnings("unused")
public class Bus {
    private static final AddonEventBus addonEventBus = new AddonEventBus();

    public static LocalAddonEventBus<?> getLocalAddonEventBus(String addonId) {
        AddonContainer container = BubbleBlaster.getInstance().getAddonManager().getContainerFromId(addonId);
        ModObject<? extends ModInstance> modObject;
        if (container != null) {
            modObject = container.getAddonObject();
            return modObject.getEventBus();
        }
        throw new NullPointerException();
    }

    public static @NotNull LocalAddonEventBus<? extends ModInstance> getLocalAddonEventBus(ModInstance addon) {
        ModObject<? extends ModInstance> modObject = addon.getAddonObject();
        return modObject.getEventBus();
    }

    public static <T extends ModInstance> LocalAddonEventBus<T> getLocalAddonEventBus(ModObject<T> modObject) {
        return modObject.getEventBus();
    }

    public static AddonEventBus getAddonEventBus() {
        return addonEventBus;
    }

    public static @NotNull GameEventBus getQBubblesEventBus() {
        return BubbleBlaster.getEventBus();
    }
}

File: src/main/kotlin/qtech/bubbles/event/Bus.kt
package qtech.bubbles.event

import qtech.bubbles.BubbleBlaster
import qtech.bubbles.mods.loader.ModContainer
import qtech.bubbles.common.mods.ModObject
import qtech.bubbles.common.mods.ModInstance
import qtech.bubbles.event.bus.AddonEventBus
import qtech.bubbles.event.bus.BBEventBus
import qtech.bubbles.event.bus.LocalAddonEventBus

object Bus {
    @kotlin.jvm.JvmStatic
    val addonEventBus = AddonEventBus()
    fun getLocalAddonEventBus(addonId: String): LocalAddonEventBus<*> {
        val container: ModContainer? = BubbleBlaster.instance.modManager.getContainer(addonId)
        val modInstanceObject: ModObject<out ModInstance> = container!!.obj
        return modInstanceObject.eventBus
    }

    fun getLocalAddonEventBus(modInstance: ModInstance): LocalAddonEventBus<out ModInstance> {
        val addonObject = modInstance.`object`
        return addonObject.eventBus
    }

    fun <T : ModInstance> getLocalAddonEventBus(modObject: ModObject<T>): LocalAddonEventBus<T> {
        return modObject.eventBus
    }

    @kotlin.jvm.JvmStatic
    val qBubblesEventBus: BBEventBus
        get() = BubbleBlaster.eventBus
}